pkg/client/client.go

```diff
@@ -201,63 +207,92 @@ func (client *DataSetClient) listenAndSendBufferForSession(session string, ch ch
+					MaxInterval:         client.Config.BufferSettings.RetryMaxInterval,
+					MaxElapsedTime:      client.Config.BufferSettings.RetryMaxElapsedTime,
+					Stop:                backoff.Stop,
+					Clock:               backoff.SystemClock,
+				}
+				expBackoff.Reset()
+				retryNum := int64(0)
+				for {
+					response, err := client.SendAddEventsBuffer(buf)
+					client.setLastError(err)
+					lastHttpStatus := uint32(0)
+					if err != nil {
+						client.Logger.Error("unable to send addEvents buffers", zap.Error(err))
+						if strings.Contains(err.Error(), "Unable to send request") {
+							lastHttpStatus = HttpErrorCannotConnect
+							client.LastHttpStatus.Store(lastHttpStatus)
+						} else {
+							lastHttpStatus = HttpErrorHasErrorMessage
+							client.LastHttpStatus.Store(lastHttpStatus)
+							client.onBufferDrop(buf, lastHttpStatus, err)
+							break
+						}
+					}
+					zaps := make([]zap.Field, 0)
+					if response.ResponseObj != nil {
+						zaps = append(
+							zaps,
+							zap.String("httpStatus", response.ResponseObj.Status),
+							zap.Int("httpCode", response.ResponseObj.StatusCode),
+						)
+						lastHttpStatus = uint32(response.ResponseObj.StatusCode)
 						client.LastHttpStatus.Store(lastHttpStatus)
-						client.buffersProcessed.Add(1)
-						continue
 					}
-				}
-				zaps := make([]zap.Field, 0)
-				if response.ResponseObj != nil {
+
 					zaps = append(
 						zaps,
-						zap.String("httpStatus", response.ResponseObj.Status),
-						zap.Int("httpCode", response.ResponseObj.StatusCode),
+						zap.String("status", response.Status),
+						zap.String("message", response.Message),
 					)
-					lastHttpStatus = uint32(response.ResponseObj.StatusCode)
-					client.LastHttpStatus.Store(lastHttpStatus)
-				}
-
-				zaps = append(
-					zaps,
-					zap.String("status", response.Status),
-					zap.String("message", response.Message),
-				)
-				client.Logger.Debug("Events were sent to DataSet",
-					zaps...,
-				)
-
-				if IsRetryableStatus(lastHttpStatus) {
-					// check whether header is specified and get its value
-					retryAfter, specified := client.getRetryAfter(
-						response.ResponseObj,
-						time.Duration(
-							int64(buf.Attempt+1),
-						)*client.Config.BufferSettings.RetryInitialInterval,
+					client.Logger.Debug("Events were sent to DataSet",
+						zaps...,
 					)
 
-					if specified {
-						// retry after is specified, we should update
-						// client state, so we do not send more requests
+					if isOkStatus(lastHttpStatus) {
+						// everything was fine, there is no need for retries
+						break
+					}
 
-						client.setRetryAfter(retryAfter)
+					backoffDelay := expBackoff.NextBackOff()
+					client.Logger.Info("Backoff + Retries: ", zap.Int64("retryNum", retryNum), zap.Duration("backoffDelay", backoffDelay))
+					if backoffDelay == backoff.Stop {
+						// throw away the batch
+						err = fmt.Errorf("max elapsed time expired %w", err)
```

Comment:
We should probably log a more explicit warning / error message when we failed to success after exhausting all the retries. Something along the lines of "Failed to send buffer after exhausting X retries (total retry / wait time of X seconds). Last error: X".